dot/telemetry/telemetry.go

```diff
@@ -100,18 +102,55 @@ func (h *Handler) SendConnection(data *ConnectionData) {
 		"version": data.SystemVersion}
 	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
 	h.telemetryLogger.Print()
-	h.sendTelemtry()
+	h.sendTelemetry()
 }
 
 // SendBlockImport sends block imported message to telemetry connection
 func (h *Handler) SendBlockImport(bestHash string, height *big.Int) {
 	payload := log.Fields{"best": bestHash, "height": height.Int64(), "msg": "block.import", "origin": "NetworkInitialSync"}
 	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
 	h.telemetryLogger.Print()
-	h.sendTelemtry()
+	h.sendTelemetry()
 }
 
-func (h *Handler) sendTelemtry() {
+// NetworkData struct to hold network data telemetry information
+type NetworkData struct {
+	Peers   int
+	RateIn  float64
+	RateOut float64
+}
+
+// SendNetworkData send network data system.interval message to telemetry connection
+func (h *Handler) SendNetworkData(data *NetworkData) {
+	payload := log.Fields{"bandwidth_download": data.RateIn, "bandwidth_upload": data.RateOut, "msg": "system.interval", "peers": data.Peers}
+	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
+	h.telemetryLogger.Print()
+	h.sendTelemetry()
+}
+
+// BlockIntervalData struct to hold data for block system.interval message
+type BlockIntervalData struct {
+	BestHash           common.Hash
+	BestHeight         *big.Int
+	FinalizedHash      common.Hash
+	FinalizedHeight    *big.Int
+	TXCount            int
+	UsedStateCacheSize int
+}
+
+// SendBlockIntervalData send block data system interval information to telemetry connection
+func (h *Handler) SendBlockIntervalData(data *BlockIntervalData) {
+	payload := log.Fields{"best": data.BestHash.String(), "finalized_hash": data.FinalizedHash.String(),
+		"finalized_height": data.FinalizedHeight, "height": data.BestHeight, "msg": "system.interval", "txcount": data.TXCount,
+		"used_state_cache_size": data.UsedStateCacheSize}
+	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
+	h.telemetryLogger.Print()
+	h.sendTelemetry()
+}
+
+func (h *Handler) sendTelemetry() {
```

Comment:
Also, the lock should be acquired in the functions calling `sendTelemetry` because `Handler.buf` is written before calling `sendTelemetry` but it is reset inside `sendTelemetry`

```
// SendBlockIntervalData send block data system interval information to telemetry connection
func (h *Handler) SendBlockIntervalData(data *BlockIntervalData) {
	h.Lock()
	defer h.Unlock()
	payload := log.Fields{"best": data.BestHash.String(), "finalized_hash": data.FinalizedHash.String(),
		"finalized_height": data.FinalizedHeight, "height": data.BestHeight, "msg": "system.interval", "txcount": data.TXCount,
		"used_state_cache_size": data.UsedStateCacheSize}
	h.telemetryLogger = log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
	h.telemetryLogger.Print()
	h.sendTelemetry()
}
```

Reply:
I've updated this to add locks and created sender which reads from buffer and sends messages.